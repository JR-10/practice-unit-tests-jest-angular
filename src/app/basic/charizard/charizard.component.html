<h1>Pokemon</h1>
<!-- se muestra si charizard no existe -->
<div *ngIf="!charizard; else elseBlock">
  <h2>Loading...</h2>
</div>


<!-- ngtemplate para colocar un fragmento de codigo  -->
<ng-template #elseBlock>
  <h3>{{charizard?.name | titlecase}}</h3>
  <img [src]="charizard?.sprites?.front_default" [alt]="charizard?.name">
</ng-template>
