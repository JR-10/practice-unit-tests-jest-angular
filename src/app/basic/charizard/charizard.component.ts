import { HttpErrorResponse } from '@angular/common/http';
import { Component, OnInit } from '@angular/core';
import { Pokemon } from '../interfaces';
import { PokemonService } from '../services/pokemon.service';

@Component({
  selector: 'app-charizard',
  templateUrl: './charizard.component.html',
  styleUrls: ['./charizard.component.scss']
})
export class CharizardComponent implements OnInit {

  charizard: Pokemon | undefined; // declarar variable chertizard
  idPokemon: number = 6;
  constructor(
    private pokemonService: PokemonService, // inyeccion del servicio
  ) { }

  ngOnInit(): void {

    // implementar el servicio dentro de OnInit
    this.pokemonService.getPokemonService(this.idPokemon).subscribe({
      next: (pokemon) => {
        this.charizard = pokemon // asignar la respuesta del servicio a la variable declarada
      },
      error: (error: HttpErrorResponse) => {
      }
    });
  }

}
