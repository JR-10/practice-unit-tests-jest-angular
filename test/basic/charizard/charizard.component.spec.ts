import { ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { CharizardComponent } from '../../../src/app/basic/charizard/charizard.component';
import { PokemonService } from '../../../src/app/basic/services/pokemon.service';
import { request } from 'http';

describe('CharizardComponent', () => {
  let component: CharizardComponent;
  let fixture: ComponentFixture<CharizardComponent>;
  let compiled: HTMLElement; // para hacer pruebas con html
  let service: PokemonService; // variable para el servicio
  let httpMock: HttpTestingController; // para inyectar informacion ficticia

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ CharizardComponent ],
      imports: [HttpClientTestingModule], // importar HttpClientTestingModule como modulo para las pruebas
      providers: [PokemonService] // importar el servicio para el test
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CharizardComponent);
    component = fixture.componentInstance;
    service = TestBed.inject(PokemonService);
    httpMock = TestBed.inject(HttpTestingController)

    fixture.detectChanges();

    compiled = fixture.nativeElement;
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe hacer un match con el snapshot', () => {
    expect(compiled.innerHTML).toMatchSnapshot();
  });

  test('debe de mostrar un loading al inicio', () => {  // prueba HTML
    const h2 = compiled.querySelector('h2'); // declara la constante
    expect (h2?.textContent ).toContain('Loading...'); // espera que dentro del h2 este el Loading...
  });

  test('debe de cargar a charizard inmediatamente', () => {  // prueba componente llamado al servicio

    // constante de la repsuesta del servicio
    const dummyPokemon = {
      name: 'charizardo!!!',
      sprites: {
        front_default: 'https://charizard.com/sprite.png',
      }
    }
    // constante para el endpoint del servicio
    const request = httpMock.expectOne('https://pokeapi.co/api/v2/pokemon/6');
    expect(request.request.method).toBe('GET'); // espera que el metodo utilizado en el request es un GET
    request.flush(dummyPokemon); // se inyecta la data, para simular que es request regreso el dummyPokemon

    fixture.detectChanges(); // dispara el ciclo de vida de Angular, regenera los componente
    // console.log(compiled.innerHTML); // validar con el console.log()

  });

});
