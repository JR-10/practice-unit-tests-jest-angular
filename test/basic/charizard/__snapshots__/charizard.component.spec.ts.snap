// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharizardComponent debe hacer un match con el snapshot 1`] = `
"<h1>Pokemon</h1><div><h2>Loading...</h2></div><!--bindings={
  "ng-reflect-ng-if-else": "[object Object]"
}--><!--container-->"
`;
